core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -341,9 +341,11 @@ class FileMessageSet private[kafka](@volatile var file: File,
     if(targetSize > originalSize || targetSize < 0)
       throw new KafkaException("Attempt to truncate log segment to " + targetSize + " bytes failed, " +
                                " size of this log segment is " + originalSize + " bytes.")
-    channel.truncate(targetSize)
-    channel.position(targetSize)
-    _size.set(targetSize)
+    if (targetSize != channel.size().toInt) {
```

Comment:
For 1:
The Javadoc of FileChannel states that "If the given size is greater than or equal to the file's current size then the file is not modified." So in my opinion an Exception is not necessary but we could change the test to targetSize < channel.size.toInt so that there will be no mtime modifications.

For 2: I will add this if 1. is clear and ready.

For 3: I tried to match the rest of the file, but I will fix this for the whole file.